src/dataflow/src/source/kafka.rs

```diff
@@ -614,27 +681,114 @@ impl ControlPlaneInfo {
             },
         );
     }
+
+    /// Generates a timestamp that is guaranteed to be monotonically increasing.
+    /// This may require multiple calls to the underlying now() system method, which is not
+    /// guaranteed to increase monotonically
+    fn generate_next_timestamp(&mut self) -> u64 {
+        // TODO(ncrooks) - If someone does something silly like sets their
+        // system clock backward by an hour while mz is running,
+        // we will hang here for an hour.
+        let mut new_ts = 0;
+        while new_ts <= self.last_closed_ts {
+            let start = SystemTime::now();
+            new_ts = start
+                .duration_since(UNIX_EPOCH)
+                .expect("Time went backwards")
+                .as_millis() as u64;
+        }
+        assert!(new_ts > self.last_closed_ts);
+        self.last_closed_ts = new_ts;
```

Comment:
I don't understand the logic here - the timestamp isn't necessarily closed yet, right?

Reply:
in RT, every timestamp is "closed" immediately.